File: src/jsc/JSONLineBuffer.rs
use bun_collections::{ByteVecExt, VecExt};
use bun_core::strings;

/// Buffer for newline-delimited data that tracks scan positions to avoid O(n²) scanning.
/// Each byte is scanned exactly once. We track:
/// - newline_pos: position of first known newline (if any)
/// - scanned_pos: how far we've scanned (bytes before this have been checked)
/// - head: offset into the buffer where unconsumed data starts (avoids copying on each consume)
///
/// When data arrives, we only scan the NEW bytes.
/// When we consume a message, we just advance `head` instead of copying.
/// Compaction only happens when head exceeds a threshold.
#[derive(Default)]
pub struct JSONLineBuffer {
    pub data: Vec<u8>,
    /// Offset into data where unconsumed content starts.
    pub head: u32,
    /// Position of a known upcoming newline relative to head, if any.
    pub newline_pos: Option<u32>,
    /// How far we've scanned for newlines relative to head.
    pub scanned_pos: u32,
    /// Sticky: a position exceeded the u32 design limit (peer buffered a
    /// >4 GiB line). The buffer is unusable; callers treat it as too-long.
    overflow: bool,
}

/// Return type of [`JSONLineBuffer::next`]: a complete message slice plus its newline offset.
pub struct Next<'a> {
    pub data: &'a [u8],
    pub newline_pos: u32,
}

impl JSONLineBuffer {
    /// Compact the buffer when head exceeds this threshold.
    const COMPACTION_THRESHOLD: u32 = 16 * 1024 * 1024; // 16 MB

    /// Get the active (unconsumed) portion of the buffer.
    fn active_slice(&self) -> &[u8] {
        &self.data.slice()[self.head as usize..]
    }

    /// Scan for newline in unscanned portion of the buffer.
    fn scan_for_newline(&mut self) {
        if self.overflow || self.newline_pos.is_some() {
            return;
        }
        let slice = self.active_slice();
        if self.scanned_pos as usize >= slice.len() {
            return;
        }

        let unscanned = &slice[self.scanned_pos as usize..];
        if let Some(local_idx) = strings::index_of_char_usize(unscanned, b'\n') {
            // `pos == u32::MAX` would overflow the `idx + 1` consumers downstream.
            match u32::try_from(self.scanned_pos as usize + local_idx) {
                Ok(pos) if pos < u32::MAX => {
                    self.newline_pos = Some(pos);
                    self.scanned_pos = pos + 1; // Only scanned up to (and including) the newline
                }
                _ => self.overflow = true,
            }
        } else {
            match u32::try_from(slice.len()) {
                Ok(n) => self.scanned_pos = n, // No newline, scanned everything
                Err(_) => self.overflow = true,
            }
        }
    }

    /// Compact the buffer by moving data to the front. Called when head exceeds threshold.
    fn compact(&mut self) {
        if self.head == 0 {
            return;
        }
        self.data.drain_front(self.head as usize);
        self.head = 0;
    }

    /// Append bytes to the buffer, scanning only new data for newline.
    pub fn append(&mut self, bytes: &[u8]) {
        let _ = self.data.write(bytes);
        self.scan_for_newline();
    }

    /// True once buffered positions exceeded the u32 design limit (a >4 GiB
    /// line); the pending message can never be decoded.
    pub fn overflowed(&self) -> bool {
        self.overflow
    }

    /// Returns the next complete message (up to and including newline) if available.
    pub fn next(&self) -> Option<Next<'_>> {
        if self.overflow {
            return None;
        }
        let pos = self.newline_pos?;
        Some(Next {
            data: &self.active_slice()[0..(pos as usize) + 1],
            newline_pos: pos,
        })
    }

    /// Consume bytes from the front of the buffer after processing a message.
    /// Just advances head offset - no copying until compaction threshold is reached.
    pub fn consume(&mut self, bytes: u32) {
        self.head = match self.head.checked_add(bytes) {
            Some(h) => h,
            None => {
                self.overflow = true;
                return;
            }
        };

        // Adjust scanned_pos (subtract consumed bytes, but don't go negative)
        self.scanned_pos = self.scanned_pos.saturating_sub(bytes);

        // Adjust newline_pos
        if let Some(pos) = self.newline_pos {
            if bytes > pos {
                // Consumed past the known newline - clear it and scan for next
                self.newline_pos = None;
                self.scan_for_newline();
            } else {
                self.newline_pos = Some(pos - bytes);
            }
        }

        // Check if we've consumed everything
        if self.head as usize >= self.data.len() {
            // Free memory if capacity exceeds threshold, otherwise just reset
            if self.data.capacity() as u32 >= Self::COMPACTION_THRESHOLD {
                self.data = Vec::<u8>::default();
            } else {
                self.data.clear();
            }
            self.head = 0;
            self.scanned_pos = 0;
            self.newline_pos = None;
            return;
        }

        // Compact if head exceeds threshold to avoid unbounded memory growth
        if self.head >= Self::COMPACTION_THRESHOLD {
            self.compact();
        }
    }

    pub fn is_empty(&self) -> bool {
        self.head as usize >= self.data.len()
    }

    /// Notify the buffer that `nread` bytes were written directly into the
    /// tail of `data` (via `data.uv_alloc_spare_u8()`).
    ///
    /// Takes a length, not a `&[u8]`, because the only caller's slice would
    /// alias `&mut self.data` — and only the length is used here. Passing the
    /// slice through would re-introduce the Stacked-Borrows hazard the
    /// `on_read` refactor removed.
    pub fn notify_written(&mut self, nread: usize) {
        // SAFETY: caller (libuv on_read) wrote `nread` bytes into the uv_alloc_spare* slice.
        unsafe { self.data.uv_commit(nread) };
        self.scan_for_newline();
    }
}

// `pub fn deinit` dropped: Vec<u8>'s Drop frees the backing allocation (global mimalloc).

File: src/jsc/JSONLineBuffer.zig
/// Buffer for newline-delimited data that tracks scan positions to avoid O(n²) scanning.
/// Each byte is scanned exactly once. We track:
/// - newline_pos: position of first known newline (if any)
/// - scanned_pos: how far we've scanned (bytes before this have been checked)
/// - head: offset into the buffer where unconsumed data starts (avoids copying on each consume)
///
/// When data arrives, we only scan the NEW bytes.
/// When we consume a message, we just advance `head` instead of copying.
/// Compaction only happens when head exceeds a threshold.
pub const JSONLineBuffer = struct {
    data: bun.ByteList = .{},
    /// Offset into data where unconsumed content starts.
    head: u32 = 0,
    /// Position of a known upcoming newline relative to head, if any.
    newline_pos: ?u32 = null,
    /// How far we've scanned for newlines relative to head.
    scanned_pos: u32 = 0,

    /// Compact the buffer when head exceeds this threshold.
    const compaction_threshold = 16 * 1024 * 1024; // 16 MB

    /// Get the active (unconsumed) portion of the buffer.
    fn activeSlice(self: *const @This()) []const u8 {
        return self.data.slice()[self.head..];
    }

    /// Scan for newline in unscanned portion of the buffer.
    fn scanForNewline(self: *@This()) void {
        if (self.newline_pos != null) return;
        const slice = self.activeSlice();
        if (self.scanned_pos >= slice.len) return;

        const unscanned = slice[self.scanned_pos..];
        if (bun.strings.indexOfChar(unscanned, '\n')) |local_idx| {
            bun.debugAssert(local_idx <= std.math.maxInt(u32));
            const pos = self.scanned_pos +| @as(u32, @intCast(local_idx));
            self.newline_pos = pos;
            self.scanned_pos = pos +| 1; // Only scanned up to (and including) the newline
        } else {
            bun.debugAssert(slice.len <= std.math.maxInt(u32));
            self.scanned_pos = @intCast(slice.len); // No newline, scanned everything
        }
    }

    /// Compact the buffer by moving data to the front. Called when head exceeds threshold.
    fn compact(self: *@This()) void {
        if (self.head == 0) return;
        const slice = self.activeSlice();
        bun.copy(u8, self.data.ptr[0..slice.len], slice);
        bun.debugAssert(slice.len <= std.math.maxInt(u32));
        self.data.len = @intCast(slice.len);
        self.head = 0;
    }

    /// Append bytes to the buffer, scanning only new data for newline.
    pub fn append(self: *@This(), bytes: []const u8) void {
        _ = bun.handleOom(self.data.write(bun.default_allocator, bytes));
        self.scanForNewline();
    }

    /// Returns the next complete message (up to and including newline) if available.
    pub fn next(self: *const @This()) ?struct { data: []const u8, newline_pos: u32 } {
        const pos = self.newline_pos orelse return null;
        return .{
            .data = self.activeSlice()[0 .. pos + 1],
            .newline_pos = pos,
        };
    }

    /// Consume bytes from the front of the buffer after processing a message.
    /// Just advances head offset - no copying until compaction threshold is reached.
    pub fn consume(self: *@This(), bytes: u32) void {
        self.head +|= bytes;

        // Adjust scanned_pos (subtract consumed bytes, but don't go negative)
        self.scanned_pos = if (bytes >= self.scanned_pos) 0 else self.scanned_pos - bytes;

        // Adjust newline_pos
        if (self.newline_pos) |pos| {
            if (bytes > pos) {
                // Consumed past the known newline - clear it and scan for next
                self.newline_pos = null;
                self.scanForNewline();
            } else {
                self.newline_pos = pos - bytes;
            }
        }

        // Check if we've consumed everything
        if (self.head >= self.data.len) {
            // Free memory if capacity exceeds threshold, otherwise just reset
            if (self.data.cap >= compaction_threshold) {
                self.data.deinit(bun.default_allocator);
                self.data = .{};
            } else {
                self.data.len = 0;
            }
            self.head = 0;
            self.scanned_pos = 0;
            self.newline_pos = null;
            return;
        }

        // Compact if head exceeds threshold to avoid unbounded memory growth
        if (self.head >= compaction_threshold) {
            self.compact();
        }
    }

    pub fn isEmpty(self: *const @This()) bool {
        return self.head >= self.data.len;
    }

    pub fn unusedCapacitySlice(self: *@This()) []u8 {
        return self.data.unusedCapacitySlice();
    }

    pub fn ensureUnusedCapacity(self: *@This(), additional: usize) void {
        bun.handleOom(self.data.ensureUnusedCapacity(bun.default_allocator, additional));
    }

    /// Notify the buffer that data was written directly (e.g., via pre-allocated slice).
    pub fn notifyWritten(self: *@This(), new_data: []const u8) void {
        bun.debugAssert(new_data.len <= std.math.maxInt(u32));
        self.data.len +|= @as(u32, @intCast(new_data.len));
        self.scanForNewline();
    }

    pub fn deinit(self: *@This()) void {
        self.data.deinit(bun.default_allocator);
    }
};

const bun = @import("bun");
const std = @import("std");
